src/internal/types.ts

```diff
@@ -118,4 +118,123 @@ export type ObservedValuesFromArray<X> = X extends Array<ObservableInput<infer T
  * If you have `T extends Array<any>`, and pass a `string[]` to it,
  * `ValueFromArray<T>` will return the actual type of `string`.
  */
-export type ValueFromArray<A> = A extends Array<infer T> ? T : never;
\ No newline at end of file
+export type ValueFromArray<A> = A extends Array<infer T> ? T : never;
+
+/** Tuple manipulation types. */
+
+export type SpecificNumber<N extends number> = number extends N ? never : N;
+export type IsSpecificNumber<N extends number> = SpecificNumber<N> extends never ? false : true;
+export type IsTuple<T extends any[]> = IsSpecificNumber<T['length']>;
+
+type Fn<Args extends any[]> = (...args: Args) => any;
+type ExtractFromArray<T extends any[]> = T extends Array<infer U> ? U : never;
+
+export type First<T extends any[]> = T[0];
+export type Rest<T extends any[]> = Fn<T> extends ((first: any, ...rest: infer R) => any) ? R : never;
+
+export type Prepend<U, T extends any[]> = ((u: U, ...t: T) => any) extends Fn<infer Pushed> ? Pushed : never;
+
+export type MakeTuple<T, Size extends number> =
+  Size extends  0 ? [] :
+  Size extends  1 ? [T] :
+  Size extends  2 ? [T, T] :
+  Size extends  3 ? [T, T, T] :
+  Size extends  4 ? [T, T, T, T] :
+  Size extends  5 ? [T, T, T, T, T] :
+  Size extends  6 ? [T, T, T, T, T, T] :
+  Size extends  7 ? [T, T, T, T, T, T, T] :
+  Size extends  8 ? [T, T, T, T, T, T, T, T] :
+  Size extends  9 ? [T, T, T, T, T, T, T, T, T] :
+  Size extends 10 ? [T, T, T, T, T, T, T, T, T, T] :
+  T[];
+
+export type Take<T extends any[], N extends number> =
+  IsTuple<T> extends false ? MakeTuple<ExtractFromArray<T>, N> :
+  Drop<T, N> extends [] ? T : // if taking more than T['length'] just return T;
+  N extends  0 ? [] :
+  N extends  1 ? [T[0]] :
+  N extends  2 ? [T[0], T[1]] :
+  N extends  3 ? [T[0], T[1], T[2]] :
+  N extends  4 ? [T[0], T[1], T[2], T[3]] :
+  N extends  5 ? [T[0], T[1], T[2], T[3], T[4]] :
+  N extends  6 ? [T[0], T[1], T[2], T[3], T[4], T[5]] :
+  N extends  7 ? [T[0], T[1], T[2], T[3], T[4], T[5], T[6]] :
+  N extends  8 ? [T[0], T[1], T[2], T[3], T[4], T[5], T[6], T[7]] :
+  N extends  9 ? [T[0], T[1], T[2], T[3], T[4], T[5], T[6], T[7], T[8]] :
+  N extends 10 ? [T[0], T[1], T[2], T[3], T[4], T[5], T[6], T[7], T[8], T[9]] :
+  T[number][];
```

Comment:
LMAO, I'm not saying this is wrong, but I am saying I'm lost with this TS usage. I'm not even really sure what this would match... `blah[number][]`? What does that match? `blah[1][2]`? wouldn't that be the same with `blah[][]`?

Reply:
Oh yeah, I can def make a tech talk on just these magic types.

Basically `tuple[number]` just creates a union of all the elements in the tuple

```ts
type MyTuple = [1, 2, 3];

let myThing: MyTuple[number];
myThing = 1;
myThing = 2;
myThing = 3;
// myThing = 4; Error
```

[demo](http://www.typescriptlang.org/play/?ssl=7&ssc=22&pln=1&pc=1#code/C4TwDgpgBAsiAqBXMAbaBeKBtAjAGigCYCBmAXQG4BYAKFrWCgFsEALASwDsBzALlgTI0WToiYAjCACdKtFvA48omHNRrzF3ZUTUauWzCTUB6Y8zb7tAFgpQAolKkB7KbVpA)

What these last cases are, are just the safe defaults to use when we can't do type magicking. In this case if we have an Observable of elements `[1, 2, 3, 4, boolean, '', 1, 2, 3, 4, 5, 6]`...

The type of this result would be `Observable<1|2|3|4|5|6|''|boolean, Array<1|2|3|4|5|6|''|boolean>>`

Reply:
Also https://github.com/kolodny/tuplizer/blob/master/index.test-d.ts has more examples of usage of these types

Reply:
Guys sorry but my eyes are bleeding so I can't understand the reason and limitation of this.

does it mean i can only emit 10 different types from observable?
like when I want to concat 10 streams into one? what will happen if i need 11?

Reply:
@TrejGun Not at all. These changes are completely backward compatible so anything you could do before this change you should be able to do after. The only thing this PR is trying to do is catch type errors on things that are probably incorrect like trying to take `6` elements on an Observable that will only ever emit 3 items. Even things like that are still possible but you need to tell the type system to ignore the number of elements in the Observable. I wish I had time to do a proper writeup the inner workings of this PR, how and why this works, how every `IsTuple<T> extends false ? ...` is a compile short circuit to speed up time the TS compiler spends on that long ternary chain, ...etc cc @benlesh on the last point

Reply:
@kolodny thanks for explanation. it would be owesome to see longer doc someday